File: linkedlist/234.c
/**
 * 快慢指针找到中点 reverse剩余一半 在余前一半遍历比较
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* reverse(struct ListNode* head);
bool isPalindrome(struct ListNode* head) {
    if(head == NULL) return true;
    struct ListNode *fast, *slow;
    fast = head;
    slow = head;
    while(fast)
    {
        fast = fast->next ? fast->next->next : NULL;
        slow = slow->next;
    }
    slow = reverse(slow);
    while(slow)
    {
        if(slow->val != head->val) return false;
        slow = slow->next;
        head = head->next;
    }
    return true;
    
}

struct ListNode* reverse(struct ListNode* head)
{
    if(head == NULL) return head;
    struct ListNode *dummy = (struct ListNode *)malloc(sizeof(struct ListNode));
    dummy->next = head;
    dummy->val = -1;
    struct ListNode *pre, *cur, *next;
    pre = dummy;
    cur = pre->next;
    while(cur->next)
    {
        next = cur->next;
        cur->next = cur->next->next;
        next->next = pre->next;
        pre->next = next;
    }
    return dummy->next;
}

File: linkedlist/234.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func isPalindrome(head *ListNode) bool {
    var index, length int
    dummy := head
    for dummy != nil {
        dummy = dummy.Next
        length++
    }

    l := make([]int, length/2)
    
    for index < length / 2 {
        l = append(l, head.Val)
        head = head.Next
        index++
    }
    if length % 2 == 1 {
        head = head.Next
        index++
    }
    for index < length {
        if head.Val != l[len(l)-1] {
            return false
        }
        index++
        head = head.Next
        l = l[:len(l)-1]
    }
    return true
}
